aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -42,383 +47,507 @@ module aptos_framework::fungible_asset {
     }
 
-    /// Assert the owner of `metadata`.
-    inline fun assert_owner<T: key>(owner: &signer, metadata: &Object<T>) {
-        assert!(object::is_owner(*metadata, signer::address_of(owner)), error::permission_denied(ENOT_OWNER));
+    fun deposit_internal<T: key>(
+        wallet: Object<T>,
+        fa: FungibleAsset,
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let FungibleAsset { metadata, amount } = fa;
+        let wallet_metadata = wallet_metadata(wallet);
+        assert!(metadata == wallet_metadata, error::invalid_argument(EFUNGIBLE_ASSET_AND_WALLET_MISMATCH));
+        let wallet_addr = object::object_address(&wallet);
+        let wallet = borrow_global_mut<FungibleAssetWallet>(wallet_addr);
+        wallet.balance = wallet.balance + amount;
+
+        let events = borrow_global_mut<FungibleAssetWalletEvents>(wallet_addr);
+        event::emit_event(&mut events.deposit_events, DepositEvent { amount });
     }
 
     /// Extract `amount` of fungible asset from `wallet`.
-    fun extract(
-        wallet: &Object<FungibleAssetWallet>,
+    fun withdraw_internal(
+        wallet_addr: address,
         amount: u64,
-    ): FungibleAsset acquires FungibleAssetWallet {
+    ): FungibleAsset acquires FungibleAssetWallet, FungibleAssetWalletEvents {
         assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
-        let wallet = borrow_fungible_asset_mut(wallet);
-        assert!(wallet.allow_ungated_transfer, error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED));
+        let wallet = borrow_global_mut<FungibleAssetWallet>(wallet_addr);
         assert!(wallet.balance >= amount, error::invalid_argument(EINSUFFICIENT_BALANCE));
         wallet.balance = wallet.balance - amount;
-        FungibleAsset {
-            metadata: wallet.metadata,
-            amount
-        }
+
+        let events = borrow_global_mut<FungibleAssetWalletEvents>(wallet_addr);
+        let metadata = wallet.metadata;
+        event::emit_event(&mut events.withdraw_events, WithdrawEvent { amount });
+
+        FungibleAsset { metadata, amount }
     }
 
     /// Increase the supply of a fungible metadata by minting.
     fun increase_supply<T: key>(metadata: &Object<T>, amount: u64) acquires FungibleAssetMetadata {
-        assert!(amount != 0, error::invalid_argument(EZERO_AMOUNT));
+        assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
         let fungible_metadata = borrow_fungible_metadata_mut(metadata);
         if (option::is_some(&fungible_metadata.maximum)) {
             let max = *option::borrow(&fungible_metadata.maximum);
-            assert!(max - fungible_metadata.supply >= amount, error::invalid_argument(ECURRENT_SUPPLY_OVERFLOW))
+            assert!(max - fungible_metadata.supply >= amount, error::invalid_argument(EMAX_SUPPLY_EXCEEDED))
         };
         fungible_metadata.supply = fungible_metadata.supply + amount;
```

Comment:
Curious, if `option::is_none(&fungible_metadata.maximum)` is true, can we ever get an overflow here on `u64::max`?

Reply:
What would overflow? The supply wouldn't be tracked anywhere

Reply:
This part has to be refactored later.